### PR TITLE
wdio-junit-reporter: Fix error with nested describes

### LINKED ISSUE #3855: junit-reporter does not allow nested describes
**Environment (please complete the following information):**
 - **WebdriverIO version: 5.7.13
 - **Mode: WDIO Testrunner
 - **If WDIO Testrunner, running sync/async:sync
 - **Node.js version: v10.15.0
 - **NPM version: 6.9.0
 - **Browser name and version: Chrome 73
 - **Platform name and version: Mac OSX
 - **Additional wdio packages used (if applicable): Dot reporter, Spec reporter

**Config of WebdriverIO**
```
const { register } = require('ts-node');
register({ files: true });
module.exports = require("./config");
```

```
import { Config, Test } from '@wdio/sync';

const config: Config = {
    specs: [
        './test/specs/**/*.spec.ts'
    ],
    exclude: [],
    maxInstances: 10,
    capabilities: [{
        maxInstances: 5,
        browserName: 'chrome'
    }],
    logLevel: 'info',
    bail: 0,
    baseUrl: 'http://localhost:4200' || process.env.baseUrl,
    waitforTimeout: 30000,
    connectionRetryTimeout: 90000,
    connectionRetryCount: 3,
    services: ['selenium-standalone'],
    framework: 'mocha',
    reporters: [
        'dot',
        'spec',
        ['junit', {
            outputDir: './junit-output'
        }],
    ],
    mochaOpts: {
        ui: 'bdd',
        timeout: 60000,
        compilers: [
            'tsconfig-paths/register'
        ]
    },
    onComplete: function(exitCode, config, capabilities, results) {
        console.log('exitCode is: ', exitCode);
        console.log('results are: ', results);
    },

}

export { config };
```

**Describe the bug**
When you run a file that contains nested describes, the spec fails without an error message.

**To Reproduce**

```
describe('first', () => {
  describe('second', () => {
    it('test', () => {
      console.log(true);
    })
  })
})
```

**Expected behavior**
Should run the test(s)/file without failures.

### PR DESCRIPTION
**Pull request by WillBrock**
## Proposed changes

Fixes #3855 

Reporters only support one spec file at a time now and when using nested describes it would not be able to find a spec file.

```
2019-04-16T17:26:57.199Z ERROR @wdio/local-runner: Failed launching test session: TypeError: Cannot read property 'replace' of undefined
    at JunitReporter.prepareXml (/home/focus/dev/trunk/focus-automation/node_modules/@wdio/junit-reporter/build/index.js:53:259)
    at JunitReporter.onRunnerEnd (/home/focus/dev/trunk/focus-automation/node_modules/@wdio/junit-reporter/build/index.js:23:22)
    at JunitReporter.WDIOReporter.on.runner (/home/focus/dev/trunk/focus-automation/node_modules/@wdio/reporter/build/index.js:181:12)
    at emitOne (events.js:116:13)
    at JunitReporter.emit (events.js:211:7)
    at reporters.forEach.reporter (/home/focus/dev/trunk/focus-automation/node_modules/@wdio/runner/build/reporter.js:53:49)
    at Array.forEach (<anonymous>)
    at BaseReporter.emit (/home/focus/dev/trunk/focus-automation/node_modules/@wdio/runner/build/reporter.js:53:20)
    at Runner.run (/home/focus/dev/trunk/focus-automation/node_modules/@wdio/runner/build/index.js:187:19)
    at <anonymous>
```

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: packages/wdio-junit-reporter/src/index.js [hunk 1]
```diff
@@ -29,7 +29,7 @@ class JunitReporter extends WDIOReporter {
             ? `${runner.sanitizedCapabilities}-${this.options.packageName}`
             : runner.sanitizedCapabilities
 
-        let index = 0
+        const index = 0
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/index.js]: I don't understand, so index will be always 0? Can we add a comment for it as to why?

**Comment by christian-bromann** [comment]
> Why we remove tests here?

Because the functionality isn't needed due to the fact that reporter have been moved to the worker process.

File: packages/wdio-junit-reporter/src/index.js [hunk 2]
```diff
@@ -40,14 +38,15 @@ class JunitReporter extends WDIOReporter {
                 continue
             }
 
-            const specFileName = runner.specs[index]
+            // The specs array will always return only one spec file since each reporter is invoked per spec file
```
Comment (BorisOsipov) [inline packages/wdio-junit-reporter/src/index.js]: yes but in spec file can be two describes, no? It was introduced in https://github.com/webdriverio/webdriverio/pull/3818

Comment (WillBrock) [reply packages/wdio-junit-reporter/src/index.js]: Hmm I'm not really sure what problem #3818 is solving. I don't see any way the same single runner instance can be invoked by multiple spec files.

Comment (BorisOsipov) [reply packages/wdio-junit-reporter/src/index.js]: spec file is one, but it has N describes like
```

describe("first", ()=> {
    before(() => {
        browser.url("http://google.com");
    });

    it('should do', () => {

    });
});

describe("second", ()=> {
    before(() => {
        browser.url("http://google.com");
    });

    it('should noo do', () => {

    });
});
```

Comment (WillBrock) [reply packages/wdio-junit-reporter/src/index.js]: I'm still not really sure what the problem is/was. It's doing the looping of the describes right above in `this.suites` so they can still have however may they want. The example above is working for me.

**Comment by WillBrock** [comment]
@naddison Do you have an example of the problem #3818 was solving?

**Comment by CrispusDH** [comment]
Do you plan to add tests?

**Comment by naddison** [comment]
> @naddison Do you have an example of the problem #3818 was solving?

3818 was solving if the runner had more than one spec file. It *shouldn't* technically happen since runners only ever have one spec file each. But it was causing a bug in an implementation I have using the junit-reporter. My change added functionality, but did not break any existing.

**Comment by WillBrock** [comment]
@CrispusDH There is not really any tests to add. 

> Did my change bring about the problem with nested describe blocks

@naddison Yes. The issue you were having could probably be addressed in another ticket.

**Comment by naddison** [comment]
So are we reverting my change? Is that's what happening?

**Comment by WillBrock** [comment]
@naddison This doesn't revert everything from your change, mainly just changes the index. I didn't see your original change when I first created the PR.

I'm fine if you can provide a fix or we revert the change. I still don't understand the problem that the original ticket was solving, from what I understand wdio will only have one spec run per runner instance.

**Comment by naddison** [comment]
Given the test case:
```javascript
describe('nested yah', () => {
    describe('navigating to contact-us page', function() {
        it('should allow user to navigating contact-us page ', function () {
          contactUs.open();     // navigating to login page
          contactUs.waitForContactPageToLoad()
          console.log(contactUs.getPageTitle());
          expect(contactUs.getPageTitle()).toContain('Contact Us')

        });
    });
});
```
The property suites `this.suites` inside the junit reporter has the following object:
```javascript
{ 'nested yahsuite1':
    SuiteStats {
      type: 'suite',
      start: 2019-04-18T16:32:46.012Z,
      _duration: 8312,
      uid: 'nested yahsuite1',
      cid: '0-0',
      title: 'nested yah',
      fullTitle: undefined,
      tests: [],
      hooks: [],
      suites: [ [Object] ],
      end: 2019-04-18T16:32:54.324Z },
   'navigating to contact-us pagesuite2':
    SuiteStats {
      type: 'suite',
      start: 2019-04-18T16:32:46.065Z,
      _duration: 8251,
      uid: 'navigating to contact-us pagesuite2',
      cid: '0-0',
      title: 'navigating to contact-us page',
      fullTitle: undefined,
      tests: [ [Object] ],
      hooks: [],
      suites: [],
      end: 2019-04-18T16:32:54.316Z } }
```

Before my change (in 5.7.8) this would output:

```xml
<testsuites>
  <testsuite name="nested_yah" timestamp="2019-04-18T09:47:48" time="7.705" tests="0" failures="0" errors="0" skipped="0">
    <properties>
      <property name="specId" value="0"/>
      <property name="suiteName" value="nested yah"/>
      <property name="capabilities" value="chrome"/>
      <property name="file" value=".\test\specs\ta-contact-us.spec.js"/>
    </properties>
  </testsuite>
  <testsuite name="navigating_to_contact_us_page" timestamp="2019-04-18T09:47:49" time="7.637" tests="1" failures="0" errors="0" skipped="0">
    <properties>
      <property name="specId" value="0"/>
      <property name="suiteName" value="navigating to contact-us page"/>
      <property name="capabilities" value="chrome"/>
      <property name="file" value=".\test\specs\ta-contact-us.spec.js"/>
    </properties>
    <testcase classname="chrome.navigating_to_contact_us_page" name="should_allow_user_to_navigating_contact_us_page" time="7.63"/>
  </testsuite>
</testsuites>
```

@WillBrock is this desired? If we are defining describe blocks as suites, the junit reporter should probably nest the test suite inside the other to reflect the hiearchy. In state of 5.7.8 (above) the junit report shows an empty test suite with no tests, and then the inner suite with the single test (it block). This seems inaccurate.

What is the desired output for junit reporter? I'm assuming we would want to aim for a valid junit xml schema.

**Comment by WillBrock** [comment]
> I propose I make the following changes:
>Document the expected output from the report in the documentation (readme)
>Document the junit schema we agree to use, a link to it perhaps.
>Change junit-reporter to output that expected output.
>Add unit tests for the the cases:
>nested describe blocks
>'xit' blocks

@naddison That sounds OK, it should probably just match what it was like in v4. I'm not familiar with the exact standard output for junit without going to look at it. 

> multiple spec files in a runner

This still is not possible in v5 of wdio though unless I'm missing something.

**Comment by naddison** [comment]
> This still is not possible in v5 of wdio though unless I'm missing something.

@WillBrock this is true, this is not possible with V5. However I have a codebase that is using wdio packages to support API testing (we really like the features of WDIO, hooks and reporters, and what not) but our implementation allows more than one spec file per instance.

I'd like to include enable junit-reporter to support multiple spec files here, but If it is undesired I can fork my own and just build it in house. I thought it would be okay to include here, as long as I'm not breaking existing functionality.

**Comment by WillBrock** [comment]
Closing in favor of #3868

**Pull request closed without merging**